### ISSUE TITLE
[no-throw-literal] false positive on generic error types

### ISSUE THREAD

**Issue by gao-sun** [typescript-eslint/typescript-eslint#3814]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-throw-literal": ["error"]
  }
}
```

```TS
const assert = <E extends Error>(value: unknown, error: E): asserts value => {
  if (!value) {
    throw error;
  }
};
```

**Expected Result**

Not reporting error.

**Actual Result**

![image](https://user-images.githubusercontent.com/14722250/131246768-cd6153d8-8fc1-4656-9144-de978d508e1c.png)

also from the command line:

`eslint path/to/assert.ts`

![image](https://user-images.githubusercontent.com/14722250/131246858-cafcecbe-57a8-42b0-8722-b8f9b84018f5.png)

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.29.3` |
| `@typescript-eslint/parser`        | `4.29.3` |
| `TypeScript`                       | `4.3.5` |
| `ESLint`                           | `7.31.0` |
| `node`                             | `14.7.1` |

**Comment by bradzacher** [comment]
This is not actually a false positive because you've actually referenced the interface `Error`, not the class `Error`:

![image](https://user-images.githubusercontent.com/7462525/131264059-2c9de3ee-2cda-441d-8df5-58511095fc95.png)

Why does this matter? Well it means that your function will fully accept an object that isn't actually an instance of `Error`:

```ts
assert(false, {message: '', name: '', stack: ''});
```

This object would (correctly) be invalid if it was inlined:
```ts
throw {message: '', name: '', stack: ''}; // Error: Expected an error object to be thrown
```

**Comment by gao-sun** [comment]
@bradzacher I tried to add `instanceof` to narrow down to the correct type, but it still reports the error.

```ts
const assert = <E extends Error>(value: unknown, error: E): asserts value => {
  if (!value) {
    if (error instanceof Error) {
      throw error; // still invalid
    }
  }
};
```

however, the code below works...

```ts
export type AssertFunction = (value: unknown, error: Error) => asserts value;

const assert: AssertError = (value, error) => {
  if (!value) {
    throw error; // pass
  }
};

assert(false, { message: '', name: '', stack: '' });
```

would mind to explain it a little bit more? thanks.

**Comment by bradzacher** [comment]
The TL;DR is that even though you've got an `instanceof` call, TS still reports the type of `error` to us as `E`. In order to get the narrowed type of the generic we would have to get the specific refinements on the generic. This can be visualised by how TS stringifies the type in intellisense, and you can also visualise the types in a tool like https://ts-ast-viewer.com/
![image](https://user-images.githubusercontent.com/7462525/131538696-3a670b6b-0270-4d1f-8bfc-dd8844869a16.png)

I did fib a little bit in that the interface itself isn't exactly the problem. The `Error` types are - to say the least - weird, and there's no way to avoid using the interface. You might expect that `Error` has a `class` defined in the types but it does not - instead it has an interface `Error`, an interface `ErrorConstructor`, and a global variable `Error: ErrorConstructor`. So there's no concrete "error base type" that you can consume here - just the interface.

That being said - using a generic there is a useless thing which just complicates your code for no real reason and is more often than not going to be wrong. DO use a generic when you need to explicitly use that generic type somewhere. DO NOT use a generic when you are just declaring a parameter.

**Comment by gao-sun** [comment]
thank you for the detailed explanation @bradzacher !
in our case, we have several customized error classes, and they all extend the base `Error` class. generic is just an abstraction to provide a better `assert()` over our packages. i'll re-think it.
